File: guardsched/planday.go
package guardsched

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"time"
)

const DefaultPlanDayStart = "05:00"

// ParsePlanDayStart parses "HH:MM" (whole hours only; minutes must be 00).
func ParsePlanDayStart(s string) (hour int, err error) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 5, nil
	}
	parts := strings.Split(s, ":")
	if len(parts) != 2 {
		return 0, fmt.Errorf("plan_day_start: expected HH:MM, got %q", s)
	}
	h, err := strconv.Atoi(strings.TrimSpace(parts[0]))
	if err != nil || h < 0 || h > 23 {
		return 0, fmt.Errorf("plan_day_start: hour must be 0–23, got %q", parts[0])
	}
	m, err := strconv.Atoi(strings.TrimSpace(parts[1]))
	if err != nil || m != 0 {
		return 0, fmt.Errorf("plan_day_start: minutes must be 00, got %q", parts[1])
	}
	return h, nil
}

// BlockStartHour returns wall-clock hour for a rotating calendar block.
func BlockStartHour(planStart int, block int, shiftHours float64) int {
	h := planStart + int(float64(block)*shiftHours)
	h %= 24
	if h < 0 {
		h += 24
	}
	return h
}

// WeekdayLongName returns English weekday name for t (UTC calendar date).
func WeekdayLongName(w int) string {
	names := []string{"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"}
	if w < 0 || w > 6 {
		return ""
	}
	return names[w]
}

// WeekdayLongNameLower returns lowercase English weekday (sunday..saturday).
func WeekdayLongNameLower(w int) string {
	n := WeekdayLongName(w)
	if n == "" {
		return ""
	}
	return strings.ToLower(n)
}

var weekdayNameToIndex = func() map[string]int {
	m := make(map[string]int)
	for i := 0; i < 7; i++ {
		m[WeekdayLongNameLower(i)] = i
	}
	return m
}()

// ParseWeekdayName maps a weekday string (case-insensitive) to 0=Sunday..6=Saturday.
func ParseWeekdayName(s string) (int, error) {
	key := strings.ToLower(strings.TrimSpace(s))
	if key == "" {
		return 0, fmt.Errorf("empty weekday name")
	}
	if i, ok := weekdayNameToIndex[key]; ok {
		return i, nil
	}
	return 0, fmt.Errorf("invalid weekday name %q (use sunday..saturday)", s)
}

// ParseDisabledWeekdays parses YAML disabled_weekdays list; returns unique sorted indices.
func ParseDisabledWeekdays(raw any) ([]int, error) {
	if raw == nil {
		return nil, nil
	}
	list, ok := raw.([]any)
	if !ok {
		return nil, fmt.Errorf("disabled_weekdays must be a list")
	}
	seen := map[int]bool{}
	var out []int
	for _, item := range list {
		i, err := ParseWeekdayName(fmt.Sprint(item))
		if err != nil {
			return nil, err
		}
		if !seen[i] {
			seen[i] = true
			out = append(out, i)
		}
	}
	sort.Ints(out)
	return out, nil
}

// WeekdayForAnchorPlanDay returns weekday index for plan day offset d from anchor (UTC calendar date at 00:00).
// Prefer WeekdayAtPlanDayStart when plan_day_start is not midnight.
func WeekdayForAnchorPlanDay(anchor time.Time, planDay int) int {
	return WeekdayAtPlanDayStart(anchor, planDay, 0)
}

// WeekdayAtPlanDayStart returns 0=Sunday..6=Saturday for the instant plan day d begins
// (anchor calendar date + d days at planStartHour UTC). Use this for disabled_weekdays so
// duties that cross midnight (e.g. 05:00–next 05:00) match the weekday at shift start.
func WeekdayAtPlanDayStart(anchor time.Time, planDay, planStartHour int) int {
	planStartHour %= 24
	if planStartHour < 0 {
		planStartHour += 24
	}
	y, m, d := anchor.Date()
	t := time.Date(y, m, d, 0, 0, 0, 0, time.UTC).AddDate(0, 0, planDay)
	t = t.Add(time.Duration(planStartHour) * time.Hour)
	return int(t.Weekday())
}

File: web/src/lib/planDay.ts
export const DEFAULT_PLAN_DAY_START = "05:00";
export const DEFAULT_PLAN_DAY_START_HOUR = 5;

export type ParsePlanDayStartResult =
  | { ok: true; hour: number; value: string }
  | { ok: false; error: string };

/** Parse "HH:MM" (whole hours only; minutes must be 00). */
export function parsePlanDayStart(raw: unknown): ParsePlanDayStartResult {
  const s = typeof raw === "string" ? raw.trim() : "";
  if (!s) {
    return { ok: true, hour: DEFAULT_PLAN_DAY_START_HOUR, value: DEFAULT_PLAN_DAY_START };
  }
  const m = /^(\d{1,2}):(\d{2})$/.exec(s);
  if (!m) {
    return { ok: false, error: "Expected HH:MM (e.g. 05:00)" };
  }
  const hour = Number(m[1]);
  const min = Number(m[2]);
  if (!Number.isFinite(hour) || hour < 0 || hour > 23) {
    return { ok: false, error: "Hour must be 0–23" };
  }
  if (min !== 0) {
    return { ok: false, error: "Minutes must be 00 (whole hours only)" };
  }
  const value = `${String(hour).padStart(2, "0")}:00`;
  return { ok: true, hour, value };
}

export function blockStartHour(planStart: number, block: number, shiftHours: number): number {
  const h = planStart + Math.round(block * shiftHours);
  return ((h % 24) + 24) % 24;
}

/** Hours from the plan-day timeline origin (0 = first plan day starts at planDayStartHour wall clock). */
export function blockTimelineOffset(
  day: number,
  block: number,
  shiftHours: number,
): number {
  return day * 24 + block * shiftHours;
}

/** Absolute timeline coordinate used by segment layout (origin = planDayStartHour). */
export function blockTimelineStartHour(
  day: number,
  block: number,
  planDayStartHour: number,
  shiftHours: number,
): number {
  return planDayStartHour + blockTimelineOffset(day, block, shiftHours);
}

/** Weekday index 0=Sunday..6=Saturday when plan day `dayIndex` begins at `planStartHour` UTC. */
export function weekdayIndexForPlanDayStart(
  anchorDate: string,
  dayIndex: number,
  planStartHour: number,
): number {
  const base = new Date(`${anchorDate}T00:00:00.000Z`);
  if (Number.isNaN(base.getTime())) return 0;
  const ps = ((Math.trunc(planStartHour) % 24) + 24) % 24;
  base.setUTCDate(base.getUTCDate() + dayIndex);
  base.setUTCHours(ps, 0, 0, 0);
  return base.getUTCDay();
}

/** UTC calendar date for plan day index (anchor YYYY-MM-DD + dayIndex). */
export function calendarDateForDay(anchorDate: string, dayIndex: number): string {
  const base = new Date(`${anchorDate}T00:00:00.000Z`);
  if (Number.isNaN(base.getTime())) return anchorDate;
  base.setUTCDate(base.getUTCDate() + dayIndex);
  return base.toISOString().slice(0, 10);
}

export function weekdayLongName(dateIso: string): string {
  const d = new Date(`${dateIso}T12:00:00.000Z`);
  if (Number.isNaN(d.getTime())) return "";
  return d.toLocaleDateString("en-US", { weekday: "long", timeZone: "UTC" });
}

/** Weekday name at plan-day start (anchor calendar date + dayIndex at planStartHour UTC). */
export function weekdayNameAtPlanDayStart(
  anchorDate: string,
  dayIndex0: number,
  planStartHour: number = DEFAULT_PLAN_DAY_START_HOUR,
): string {
  const wd = weekdayIndexForPlanDayStart(anchorDate, dayIndex0, planStartHour);
  const names = ["Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"];
  const calendarDate = calendarDateForDay(anchorDate, dayIndex0);
  return names[wd] ?? weekdayLongName(calendarDate);
}

/** Weekday at plan-day start for a single calendar date (plan day 0 = that date). */
export function weekdayNameForCalendarPlanDay(
  calendarDate: string,
  planStartHour: number = DEFAULT_PLAN_DAY_START_HOUR,
): string {
  return weekdayNameAtPlanDayStart(calendarDate, 0, planStartHour);
}

export function planDayTitle(
  dayIndex0: number,
  anchorDate: string,
  planStartHour: number = DEFAULT_PLAN_DAY_START_HOUR,
): string {
  const calendarDate = calendarDateForDay(anchorDate, dayIndex0);
  const weekday = weekdayNameAtPlanDayStart(anchorDate, dayIndex0, planStartHour);
  return `Day ${dayIndex0 + 1} — ${weekday}, ${calendarDate}`;
}

export function planDayTitleForCalendarDate(
  dayIndex0: number,
  calendarDate: string,
  planStartHour: number = DEFAULT_PLAN_DAY_START_HOUR,
): string {
  const weekday = weekdayNameForCalendarPlanDay(calendarDate, planStartHour);
  return `Day ${dayIndex0 + 1} — ${weekday}, ${calendarDate}`;
}

export function resolvePlanDayStartHour(
  meta?: Record<string, unknown> | null,
  globalPlanDayStart?: unknown,
): number {
  if (meta != null) {
    const mh = meta.plan_day_start_hour;
    if (typeof mh === "number" && Number.isFinite(mh)) {
      return ((Math.trunc(mh) % 24) + 24) % 24;
    }
    const parsed = parsePlanDayStart(meta.plan_day_start);
    if (parsed.ok) return parsed.hour;
  }
  const parsed = parsePlanDayStart(globalPlanDayStart ?? DEFAULT_PLAN_DAY_START);
  return parsed.ok ? parsed.hour : DEFAULT_PLAN_DAY_START_HOUR;
}

export function formatWallClockHour(hour: number): string {
  const h = ((Math.trunc(hour) % 24) + 24) % 24;
  return `${String(h).padStart(2, "0")}:00`;
}

/** X-axis layout for soldier timeline charts (0% = plan day start, not midnight). */
export function timelineChartLayout(
  days: number,
  planDayStartHour: number,
  totalSpanHours?: number,
) {
  const spanHours = totalSpanHours ?? days * 24;
  const origin = ((Math.trunc(planDayStartHour) % 24) + 24) % 24;
  return {
    spanHours,
    origin,
    segmentLeftPct(startHour: number): number {
      return ((startHour - origin) / spanHours) * 100;
    },
    segmentWidthPct(duration: number): number {
      return (duration / spanHours) * 100;
    },
    dayBoundaryLeftPct(afterDayIndex: number): number {
      return ((afterDayIndex + 1) * 24) / spanHours * 100;
    },
    tickLeftPct(offsetFromOrigin: number): number {
      return (offsetFromOrigin / spanHours) * 100;
    },
    tickWallClock(offsetFromOrigin: number): string {
      return formatWallClockHour(origin + offsetFromOrigin);
    },
  };
}

export function resolvePlanDayStartString(
  meta?: Record<string, unknown> | null,
  globalPlanDayStart?: unknown,
): string {
  if (meta != null && typeof meta.plan_day_start === "string") {
    const p = parsePlanDayStart(meta.plan_day_start);
    if (p.ok) return p.value;
  }
  const p = parsePlanDayStart(globalPlanDayStart ?? DEFAULT_PLAN_DAY_START);
  return p.ok ? p.value : DEFAULT_PLAN_DAY_START;
}
